[Step3] 데이터베이스 적용 완료

LeeSwallow opened the pull request:
CustomPage 객체를 구현해 JPA 없이 페이지네이션을 구현해보았습니다.
CustomPage 또한 Record로 변환하여 Getter 사용을 최소화 하였습니다.

### 질문사항
1. errorMessageRespose, CustomPage를  record로 변경하는 중 사용자 정의 생성자를 사용하지 못해 builder로 대체하였는데 제대로된 설계인지 모르겠습니다.

LeeSwallow committed:
feat: 페이지네이션 구현 완료

- product-list.html 에 페이지 선택 li 구현 완료
- CustomPage 구현을 통해(JPA 대신) Page dto 구성 완료
- controller & service 단에서 pagination 처리 + 기능 도입 완료

LeeSwallow committed:
fix: 불필요한 속성 삭제

LeeSwallow committed:
refactor: dto를 record 형식으로 변경

LeeSwallow committed:
refactor: dto를 record 형식으로 변경

LeeSwallow committed:
refactor: deleteById의 이중 책임 문제 해결

LeeSwallow committed:
refactor: CustomerPage Record 형식으로 대체

kcm0147 commented:
@LeeSwallow 안녕하세요 민형님! 페이지네이션 요구사항은 Step2에 없던 건데 완료하셨군요 👍 
다만, Step2 리뷰는 현재 끝난 상태이고 해당 커밋을 Step3에 반영 해 주시면 같이 리뷰해드릴게요! 해당 PR은 Close 할 필요 없이 제목만 바꾸시고 Step3를 구현하셔도 됩니다 😄

LeeSwallow commented:
감사합니다!

Commit:
refactor: Update 로직 service와 repository에 맞게 명확하게 분리되도록 분리

Commit:
docs: 3단계 구현사항 정리

Commit:
feat: 데이터베이스 설정 + sql 스크립트 작성

+ init-db.sql을 통해 product 테이블을 작성하였습니다.
+ h2 test db를 생성하고 init-db로 초기화하는 로직을 설정에 추가하였습니다.

Commit:
feat: sql 내부로직으로 인한 에러 발생 시 던질 에러와 처리할 로직 구현

+ GlobalExceptionHandler에서 해당 에러를 500 code로 처리합니다.

Commit:
feat: dbcClient 이용한 상품 데이터베이스 CRUD 구현

Commit:
feat: repository와 dao 연동

Commit:
fix: save를 통해 update를 적용하도록 재변경

Commit:
fix: 구현 완료 표시 & 상품 조회 API 수정

LeeSwallow commented:
### 진행 관련 애로사항 정리

1. 예외 처리(https://github.com/next-step/spring-gift-product/pull/138/commits/b737712315cbd5de6c2d81fc5f28dedde5c63867)

    + `JdbcClient`를 통해 구현하였고 SQL 에 대한 에러 처리를 `DBServerException`이라는 exception을 만들어 처리하였습니다.
    
    + `GlobalExceptionAdvice`에서 해당 에러를 intenal error로 처리해서 핸들링해줬습니다.
    
    + 가능한 internal error가 일어나는 일은 없어야 한다는 점을 이해하였지만 sql 처리 과정에서 일어난 문제에 대한 적절한 처리라고 생각했습니다.

2. dao와 repository(https://github.com/next-step/spring-gift-product/pull/138/commits/3fcf155d678e0eb1af1084521d28f1c6178c9d51)

    + dao에서는 db에 접근하는 밀접한 로직 처리 repository에서는 추상화된 저장소 관련 로직으로 이해하였습니다.
    
    + 따라서 dao 에서 반환하는 값들은 `JdbcClient`의 원 함수 호출 결과(영향으로 받은 row 수 반환)와 가깝게 설계하였습니다.
    
    + repository에서 제대로 적용되었는지(rownum 기반) 확인하고 처리 혹은 처리된 객체(Product) 를 다시 findById로 반환하는 식으로 처리했습니다.
    
    + 관련된 대부분의 error 처리를 repository에서 구현했습니다.


### 의문사항

1. `CustmerPage`의 Builder 설계 관련

    + Pagination은 Service 단에서 모든 객체를 받은 뒤 관련 정보를 처리하는 식으로 설계하였습니다. ([1911561](https://github.com/next-step/spring-gift-product/pull/138/commits/1911561897f9e0997f91b6bdec7df3de8f343620)), 이를 repository단으로 옮기고 repository에 대응되어 CustomPage를 생성하는 Builder 로직을 따로 구현해야 할까요?
    
1. Dto 남용 관련
    + `CustomerPage`를 dto 패키지에 저장하는 것이 옳을까요?
    + 다른 request, response도 Controller에서 검증 후 그대로 넘겨주어,  service 단에서 dto로 요청받아 응답하는 식으로 처리하고 싶은 유혹이 들곤 합니다. 결합도 측면에선 분명히 안좋은 설계이지만, 모든 값을 파라미터로 전달할때의 가독성, 중복설계 문제가 심해지는 것 같아 개인적인 의견을 듣고 싶습니다.

kcm0147 commented on src/main/java/gift/repository/ProductRepositoryImpl.java:
```diff
@@ -48,11 +48,36 @@ public Product save(Product product) {
+            Product product = productMap.get(productId);
+            product.setName(name);
+            return product;
+        }
```
이 부분이 updatePrice, updateImageUrl과 같이 중복되는 부분이 많은데, 어떻게 줄일 수 있을지 생각을 해 보셨을 것 같아요.
`함수형 인터페이스`라는 것이 존재하는데, 이를 활용하면 다음과 같이 중복을 제거 할 수 있을 것 같아요.
이런게 있다는 정도만 알고 넘어가도 좋아요.

```suggestion
private Product updateFieldById(Long productId, Consumer<Product> updater) {
    Product product = productMap.get(productId);
    if (product != null) {
        updater.accept(product);
    }
    return product;
}

public Product updateNameById(Long productId, String name) {
    return updateFieldById(productId, product -> product.setName(name));
}

@Override
public Product updatePriceById(Long productId, Long price) {
    return updateFieldById(productId, product -> product.setPrice(price));
}

@Override
public Product updateImageUrlById(Long productId, String imageUrl) {
    return updateFieldById(productId, product -> product.setImageUrl(imageUrl));
}
```

ref. https://inpa.tistory.com/entry/%E2%98%95-%ED%95%A8%EC%88%98%ED%98%95-%EC%9D%B8%ED%84%B0%ED%8E%98%EC%9D%B4%EC%8A%A4-API

LeeSwallow replied:
"PATCH와 같이 일부를 수정하기 위한 기능들도 만들어야 하지 않을까?" 라는 생각에서 각 필드에 대한 개별 구현을 만들었습니다. 

 updater로 권한을 위임한다면 바뀐 Product의 내용을 확인할 수 있어야 dao 단에서 개별 적용가능한데  그러기 위한 비용이 너무 커져서 어떻게 할 지 의문입니다.

kcm0147 commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -54,20 +54,26 @@ public Product create(Product product) {
-        if (product.getImageUrl() != null) existingProduct.setImageUrl(product.getImageUrl());
 
-        return productRepository.update(existingProduct);
+        if (product.getName() != null)
```
Java에서는 중괄호를 생략할 수 있지만, 버그 방지를 위해서 보통 중괄호를 사용하는 것을 권장하고있어요.
Goole Style Guide에도 **"Braces are used with if, else, for, do, and while statements, even when the body is empty or contains only a single statement."** 라고 적혀있습니다.
중괄호를 사용하는 습관을 들이면 좋을 것 같아요.

kcm0147 commented on src/main/java/gift/exception/DBServerException.java:
```diff
@@ -0,0 +1,10 @@
+package gift.exception;
```
기존 JDBC, JPA에서 제공하는 여러 Exception들이 있을텐데요 혹시 DBServerException을 따로 커스텀하게 정의한 이유가 있을까요?

LeeSwallow replied:
checked exception인 SQLEception을 제외하고는 대응되는 에러가 생각나지 않았습니다..

kcm0147 commented on document/상품 조회 API 명세서.md:
```diff
@@ -10,3 +10,199 @@
 | `/api/products/{productId}` | PUT    | 단건 상품 수정 | 특정 상품의 정보를 수정합니다. |
 | `/api/products/{productId}` | DELETE | 단건 상품 삭제 | 특정 상품을 삭제합니다.     |
 
+## 요청 및 응답 예시
```
명세를 매우 자세하게 써 주셨군요.. 💯

kcm0147 commented on src/main/java/gift/dao/ProductDao.java:
```diff
@@ -0,0 +1,106 @@
+
+@Repository
+public class ProductDao {
+    private final JdbcClient jdbcClient;
```
JDBC Client를 이용해서 구현하셨네요 👍

kcm0147 commented on src/main/java/gift/dao/ProductDao.java:
```diff
@@ -0,0 +1,106 @@
+                .toList();
+    }
+
+    public Product findById(Long productId) {
+        String sql = "SELECT * FROM products WHERE id = ?";
+        return jdbcClient.sql(sql)
+                .param(productId)
+                .query(new ProductRowMapper())
+                .stream()
+                .findFirst()
+                .orElse(null);
+    }
```
findByID를 할 때, 결과가 없으면 orElse(null)로 반환하고있는데, Optional을 사용해서 좀 더 안전하게 처리하는 건 어떨까요?

```suggestion
public Optional<Product> findById(Long productId) {
    String sql = "SELECT * FROM products WHERE id = ?";
    return jdbcClient.sql(sql)
        .param(productId)
        .query(productRowMapper)
        .stream()
        .findFirst();
}
```

kcm0147 commented on src/main/java/gift/dao/ProductDao.java:
```diff
@@ -0,0 +1,106 @@
+            .param(product.getPrice())
+            .param(product.getImageUrl())
+            .update(keyHolder);
+        if (keyHolder.getKey() == null) {
+            throw new DBServerException("상품을 저장하는 중 오류가 발생했습니다.(key에 제대로 생성되지 않았음)");
+        }
```
예외처리를 하려고 했던 점은 너무 좋은 것 같습니다.
다만 keyHolder.getKey() == null 인 부분이, "DB 시스템 상의 문제"인건지, keyHolder에서 Key를 설정하지 못한 건지 등 정확한 이유를 알 수 없을 것 같다는 생각이 들어요.

예를 들어,
```suggestion
public Long insertWithKey(Product product) {
    KeyHolder keyHolder = new GeneratedKeyHolder();
    String sql = "INSERT INTO products (name, price, image_url) VALUES (?, ?, ?)";

    try {
        jdbcClient.sql(sql)
            .param(product.getName())
            .param(product.getPrice())
            .param(product.getImageUrl())
            .update(keyHolder);
    } catch (Exception e) {
        throw new DBServerException("DB Insert 중 예외가 발생했습니다.", e); // DB 서비스에서 발생한 예외라면 정확한 예외를 담아서 전파하는 게 좋기 때문에 exception을 추가해주는 게 좋아요.
    }

    Number key = keyHolder.getKey();
    if (key == null) {
        throw new DBServerException(String.format("상품 저장은 성공했지만, 생성된 키가 반환되지 않았습니다. 상품 정보: %s", product));
    }
...
}

```

kcm0147 commented on src/main/java/gift/dao/ProductDao.java:
```diff
@@ -0,0 +1,106 @@
+                .update();
+    }
+
+    public int updateNameById(Long productId, String name) {
+        String sql = "UPDATE products SET name = ? WHERE id = ?";
+        return jdbcClient.sql(sql)
+                .param(name)
+                .param(productId)
+                .update();
+    }
+
+    public int updatePriceById(Long productId, Long price) {
+        String sql = "UPDATE products SET price = ? WHERE id = ?";
+        return jdbcClient.sql(sql)
+                .param(price)
+                .param(productId)
+                .update();
+    }
+
+    public int updateImageUrlById(Long productId, String imageUrl) {
+        String sql = "UPDATE products SET image_url = ? WHERE id = ?";
+        return jdbcClient.sql(sql)
+                .param(imageUrl)
+                .param(productId)
+                .update();
+    }
```
이 부분도 중복을 제거할 수 있는 방법이 있는지 고민해 주세요 :)

kcm0147 commented on src/main/java/gift/dto/ErrorMessageResponse.java:
```diff
@@ -3,26 +3,45 @@
-    private final int status;
-    private final String error;
-    private final String path;
+public record ErrorMessageResponse (
```
ErrorMessageResponse의 경우 빌더패턴을 도입한 이유가 있을까요?
빌더 패턴은 선택적으로 필드를 넣어야할 경우에 쓰면 좋은데, ErrorReponse의 경우 StackTrace 필드만 선택적인 경우일 것 같아요.

LeeSwallow replied:
class => reocord로 변경 후 커스텀한 매개변수에 대한 생성자를 고민하던 중 ,
ide 메뉴에 "builder 만들기"와 같은 메뉴가 발견했던걸로 기억합니다.

kcm0147 commented on src/main/java/gift/dto/CustomPage.java:
```diff
@@ -0,0 +1,55 @@
+                throw new IllegalArgumentException("페이지 번호는 0 이상이어야 하고, 크기는 1 이상이어야 합니다!");
+            }
+            int totalPages = (int) Math.ceil((double) totalElements / size);
+            if (this.page >= totalPages) {
+                throw new IllegalArgumentException("페이지 번호가 총 페이지 수를 초과했습니다!");
+            }
```
total page가 0과 this.page가 0일 경우, 즉 아무 content가 없는 경우에 해당 에러가 발생할 것 같은데 어떻게 생각하시나요?!

kcm0147 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -15,17 +16,29 @@ public class GlobalExceptionHandler {
-        return new ResponseEntity<>(new ErrorMessageResponse(
-                request, e.getMessage(), HttpStatus.NOT_FOUND
-        ), HttpStatus.NOT_FOUND);
+        var errorMessage = new ErrorMessageResponse.Builder(request, e.getMessage(), HttpStatus.NOT_FOUND)
+                .stackTrace(e.getStackTrace() != null ? e.getStackTrace()[0].toString() : "")
+                .build();
+        return new ResponseEntity<>(errorMessage, HttpStatus.NOT_FOUND);
```
이 함수를 추출할 수 있을 것 같다는 생각이 드네요!

kcm0147 commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -22,6 +25,17 @@ public List<Product> getAll() {
     }
 
+    @Override
+    public CustomPage<Product> getBy(int page, int size) {
+        List<Product> products = productRepository.findAll().stream()
+                .sorted(Comparator.comparing(Product::getId))
+                .collect(Collectors.toList());
+
+        return new CustomPage.Builder<>(products)
+                .page(page).size(size)
+                .build();
+    }
+
```
아하.. 페이지네이션을 구현하셨다고 했는데, 전체 결과를 조회하고 애플리케이션 단에서 size만큼 데이터를 반환하도록 구현하셨군요.. 페이지네이션의 장점이 없는 구현이네요 😢

LeeSwallow replied:
jpa 추가 시 implementaion을 목적으로 관리자 페이지를 위해 임시 도입하였습니다 
page 객체 도입 이후에 from page화 같은 mapper 함수를 CusomPage dto에 추가하고 service 레이어에서 변환할 계획이었지만
페이지네이션 구현 이란 주제로 내세울 내용은 아니었던것 같습니다..

2단계에 반영하여 sql을 통한 제대로된 pagination을 구현해보겠습니다

kcm0147 reviewed (approved):
민형님 3단계도 매우 잘 구현 해 주셨어요. Builder 패턴도 쓰려고 하셨고, DB System에서 발생한 Exception 처리를 하기 위해서 CustomException도 잘 구현 해 주셨어요.
다만 요구사항에는 없었던 Pagenation은 수정이 필요한 상황인데요, 요구사항이 아니기 때문에 감안하고 머지하겠습니다 :)

---
1. CustomPage의 경우 별도로 만드신 이유가, 특정 DB API에서 제공하는 Page 모델에 의존하지 않기 위해서 별도로 만들었다고 이해했습니다. 하지만 Pagenation의 큰 장점 중 하나로 "전체 데이터"를 로딩하지않고 일부의 데이터만 로딩을 해야하는데, 현재는 전체 데이터를 로딩하고 있어서,  Pagenatinon을 구현했다고 보기엔 힘드네요.. 
```java
public List<Product> findByPage(int page, int size) {
    String sql = "SELECT * FROM products ORDER BY id LIMIT ? OFFSET ?";
    return jdbcClient.sql(sql)
        .param(size)
        .param(page * size)
        .query(productRowMapper)
        .toList();
}

or

public int countAll() {
    String sql = "SELECT COUNT(*) FROM products";
    return jdbcClient.sql(sql)
        .query(Integer.class)
        .single();
}

```
Repository Layer에서 이런 함수를 이용해서 Pagenation을 구현하면 좋을 것 같아요.

2. Dto 남용과 관련된 질문을 남겨주셨는데요, 정답은 아니지만 CustomPage는 Repository, Service , Controller Layer에서 전부 사용하기 때문에 Model 패키지에 있어도 상관없다고 생각합니다.

그리고 DTO를 Service Layer까지 사용해도 되냐 안되냐는 끝없는 논쟁이라서.. 정답은 없다고 생각합니다.
저는 다음과 같은 생각을 가지고 있어요.

**Service는 '도메인 로직 실행을 위한 명령을 받는 계층'이고,
Controller는 '외부 입력을 해석하고 내부에 전달하는 역할'**

저는 ServiceLayer로 넘기기전에 DTO는 Domain 모델로 변경해서 넘기고, 반환 값도 DTO로 반환하지않고 Domain 모델을 반환하고 Controller에서 Resopnse 모델로 변환해서 반환합니다.

그 이유로는, 도메인 로직과 무관한 외부 계층 변경이 Service에 영향을 주고(API 스펙 변경 → Service 메서드 시그니처 변경), 테스트 시에도 DTO 객체를 만들어야 하는 이상한 점이 있다고 생각해요.

---
민형님이 작성하신 코드를 보면 고민을 되게 많이 했을 것 같다는 생각이 드네요 😄 
제가 민형님께 한 리뷰는 끝이지만, 궁금한 점이 있으면 언제든 DM으로 연락주세요. 
너무 고생많으셨습니다 👍
